Assets/HoloToolkit/SpatialMapping/Scripts/SpatialMappingObserver.cs

```diff
@@ -225,80 +247,62 @@ public bool SetObserverOrigin(Vector3 origin)
         /// <param name="elapsedCookTimeSeconds">Seconds between mesh cook request and propagation of this event.</param>
         private void SurfaceObserver_OnDataReady(SurfaceData cookedData, bool outputWritten, float elapsedCookTimeSeconds)
         {
-            //We have new visuals, so we can disable and cleanup the older surface
-            GameObject surfaceToCleanup;
-            if (pendingCleanup.TryGetValue(cookedData.id.handle, out surfaceToCleanup))
+            if (outstandingMeshRequest == null)
             {
-                CleanupSurface(surfaceToCleanup);
-                pendingCleanup.Remove(cookedData.id.handle);
-            }
+                Debug.LogError(string.Format("Got OnDataReady for surface {0} while no request was outstanding.",
+                    cookedData.id.handle
+                    ));
 
-            GameObject surface;
-            if (surfaces.TryGetValue(cookedData.id.handle, out surface))
-            {
-                // Set the draw material for the renderer.
-                MeshRenderer meshRenderer = surface.GetComponent<MeshRenderer>();
-                meshRenderer.sharedMaterial = SpatialMappingManager.Instance.SurfaceMaterial;
-                meshRenderer.enabled = SpatialMappingManager.Instance.DrawVisualMeshes;
-
-                if (SpatialMappingManager.Instance.CastShadows == false)
-                {
-                    meshRenderer.shadowCastingMode = UnityEngine.Rendering.ShadowCastingMode.Off;
-                }
+                return;
             }
 
-            surfaceWorkOutstanding = false;
-            SurfaceObserver.SurfaceDataReadyDelegate dataReady = DataReady;
-            if (dataReady != null)
+            if (false
+                || (outstandingMeshRequest.Value.ID != cookedData.id.handle)
+                || (outstandingMeshRequest.Value.Filter != cookedData.outputMesh)
+                || (outstandingMeshRequest.Value.Collider != cookedData.outputCollider)
+                )
```

Comment:
I pulled the condition out into a helper function and think it works pretty well.  Thanks!  :)